Documentation for GPO trigger and stimulator reports

cjsha opened the pull request:
Fix #286 Fix #287

cjsha committed:
Documentation for GPO trigger and stimulator reports

cjsha committed:
Trigger GPO key using X key instead of square key

- Add note GPO trigger on example workflow page

bparks13 commented on articles/hardware/hs64/gpo-trigger.md:
```diff
@@ -0,0 +1,49 @@
+trigger electrical stimulus, the electrical stimulator should be the only
+stimulator device armed on the headstage. If you want to use the GPO trigger to
+only trigger optical stimulus, the optical stimulator should be the only
+stimulator device armed on the headstage. 
```
Should we reword this and put it in a caution/warning admonition? This is important, and I worry it will get lost in the paragraph as is

cjsha replied:
I made it a note. If it looks good, please resolve convo

cjsha replied:
I just resolved this convo bc we have a related ongoing conversation in a different thread [here](https://github.com/open-ephys/bonsai-onix1-docs/pull/289#discussion_r2481558356).

bparks13 commented on articles/hardware/hs64/workflow.md:
```diff
@@ -10,6 +10,7 @@ The example workflow below can by copy/pasted into the Bonsai editor using the c
 - Automatically commutates the tether if there is a proper commutator connection. 
 - Applies electrical stimulation triggered by pressing the breakout board's △ key.
 - Applies optical stimulation triggered by pressing the breakout board's ◯ key.
+- Applies both stimulation types with a lower latency trigger mechanism by pressing the breakout board's X key.
```
I would say "either" instead of "both". Although, technically, saying "both" is correct we want to highlight that it should only be used for one or the other

cjsha replied:
Agreed, I will edit this. If you think it's better, please resolve.

cjsha commented on articles/hardware/hs64/estim.md:
```diff
@@ -3,26 +3,42 @@ uid: hs64_estim
-`DeviceName` property is set to "BreakoutBoard/DigitalInput". This links the `DigitalInput` operator
-to the corresponding configuration operator. 
+[DigitalInputDataFrames](xref:OpenEphys.Onix1.DigitalInputDataFrame). Although
+the digital inputs are sampled at 4 Mhz, these data frames are only emitted when
```
yeah I describe that in the breakout board example workflow. I'll be more explicit that this is only true when that property is left blank.

cjsha committed:
Address bparks feedback

bparks13 commented on articles/hardware/hs64/estim.md:
```diff
@@ -3,26 +3,36 @@ uid: hs64_estim
+`HasFlags` operator, the enumerated values appear in the `HasFlags`'s `Value`
+property's dropdown menu. Because `HasFlags`'s `Value` is set to "Triangle", its
+output is "True" when the selected `BreakoutButtonState` bit field contains the
+"Triangle" flag. The <xref:Bonsai.Reactive.DistinctUntilChanged> operator only
```
This paragraph also seems to be duplicated in the ostim.md page, should we add this as a separate Include?

cjsha replied:
I didn't put this part in the "include" because there is a small difference between them - one is triggered using Triangle button and the other is triggered using Circle button, I believe. Everything else is the same though. I suppose I could use an "include" for everything before and another "include" for everything after, but I'm concerned about annoying others if they wanted to read the raw markdown files and edit them.. What do you think?

bparks13 commented on articles/hardware/hs64/gpo-trigger.md:
```diff
@@ -0,0 +1,43 @@
+> electrical stimulator should be the only stimulator device armed on the
+> headstage. If you want to use the GPO trigger to only trigger optical
+> stimulus, the optical stimulator should be the only stimulator device armed on
+> the headstage. 
```
I like the note, but I think this would be better served by reducing the length. These two sentences can be boiled down to this: "When using the GPO trigger, ensure that only one stimulator (electrical or optical) is armed on the headstage." We might need to add some more detail in the paragraph above that the GPO trigger will trigger both stimulators if they are both armed, but I believe that keeping the admonitions short will help drive the point home better.

cjsha replied:
I think you're right. I just learned if both are enabled and armed, the estim supercedes ostim anyway. Btw, from talking to jon just now, I think that estim supercedes ostim if they are both armed.

bparks13 commented on articles/hardware/hs64/ostim.md:
```diff
@@ -3,26 +3,36 @@ uid: hs64_ostim
+`HasFlags` operator, the enumerated values appear in the `HasFlags`'s `Value`
+property's dropdown menu. Because `HasFlags`'s `Value` is set to "Circle", its
+output is "True" when the selected `BreakoutButtonState` bit field contains the
+"Circle" flag. The <xref:Bonsai.Reactive.DistinctUntilChanged> operator only
```
I guess the value here ("Circle") is different than the value in the estim.md page. Can we make this paragraph a template instead of an include so we can modify this value? Or is it now getting to be too much effort to maintain since this only appears twice?

bparks13 commented on articles/hardware/hs64/stimulator-data.md:
```diff
@@ -0,0 +1,38 @@
+which contain data about when an electrical stimulus was delivered and the
+corresponding electrical stimulation waveform. A frame is emitted when an
+electrical stimulus is delivered. In the Headstage 64 example workflow, the
+`Headstage64ElectricalStimulatorData`'s `DeviceName` property is set to
```
Combine using dot notation?

bparks13 reviewed (commented):
It's looking really good! I've got a few more comments, but once those are addressed I am ready to merge.

cjsha commented on articles/hardware/breakout/digital-inputs.md:
```diff
@@ -14,15 +14,16 @@ functionality by responding to button presses and saves digital inputs data.
-Board use 3.3V logic levels but are also 5V tolerant. In the Breakout
-Board example workflow, the `DigitalInput`'s `DeviceName` property is set to
+Board use 3.3V logic levels but are also 5V tolerant. In the Breakout Board
+example workflow, the `DigitalInput`'s `DeviceName` property is set to
```
I think this is a valid concern, but the problem persists across the docs not just for the pages that are being added/edited in this PR. I think it's worth a separate issue  (#291) and we can resolve this convo in the meanwhile

bparks13 commented on workflows/hardware/hs64/load-hs64.py:
that makes sense, we can leave it for now.

cjsha committed:
bparks feedback, second pass

Also:
- Add dotnet xrefmap in docfx.json so I can use `<xref:System.Double>`.
- Remove include for now, the long-term solution will be to add a template.
- Remove `Condition` operator from Headstage64GpoTrigger branch
  This node's unnecessary bc only True values trigger stimulus anyway.

cjsha commented:
Note to self: make sure the stimuli waveforms have reasonable parameters.

bparks13 reviewed (approved):
I like the changes! I see some other comments, so I won't merge this, but feel free to merge it if we want to address the comments in other PRs

jonnew commented on articles/hardware/hs64/stimulator-data.md:
```diff
@@ -0,0 +1,38 @@
+"Headstage64/Headstage64ElectricalStimulator". This links the
+`Headstage64ElectricalStimulatorData` operator to the corresponding
+configuration operator. Frames from this operators are saved to a file named
+"estim_.csv" using a <xref:Bonsai.IO.CsvWriter>.
```
estim_\<timestamp\>.csv ?

cjsha replied:
`estim_<filecount>.csv`

jonnew commented on articles/hardware/hs64/workflow.md:
```diff
@@ -10,6 +10,9 @@ The example workflow below can by copy/pasted into the Bonsai editor using the c
 - Automatically commutates the tether if there is a proper commutator connection. 
 - Applies electrical stimulation triggered by pressing the breakout board's △ key.
 - Applies optical stimulation triggered by pressing the breakout board's ◯ key.
+- Applies either electrical or optical stimulation (depending on which stimulators
```
Just armed, enable determines if they produce frames.

jonnew reviewed (changes requested):
A few comments. Looks good.

cjsha commented on workflows/hardware/hs64/hs64.bonsai:
```diff
@@ -87,7 +87,7 @@
             <onix1:DeviceAddress>256</onix1:DeviceAddress>
             <onix1:Enable>true</onix1:Enable>
-            <onix1:DspCutoff>Off</onix1:DspCutoff>
+            <onix1:DspCutoff>Dsp146mHz</onix1:DspCutoff>
```
Is it worth changing the default value in the node too?

```
        public Rhd2164DspCutoff DspCutoff { get; set; } = Rhd2164DspCutoff.Dsp146mHz;
```

https://github.com/open-ephys/bonsai-onix1/blob/7fb9332eb67c111a894de5ecaddbe63def120562/OpenEphys.Onix1/ConfigureRhd2164.cs#L55

If so, I can spawn an issue in the bonsai-onix1 repo before resolving this.

jonnew replied:
Yes! Good catch

cjsha committed:
Address PR feedback

Also address jon's oral feedback, including changes to the wf and loading script:
- use spikeinterface
- plot stim data
- offset plotted channels & add scale bar
- separate ephys plots from other plots
-fix aux channel conversion
